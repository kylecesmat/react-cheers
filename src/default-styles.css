.notification-composer {
  position: absolute;
  top: 0;
  right: 0;
  width: 100%;
  max-width: 300px;
  display: flex;
  flex-direction: column;
}

.notification {
  position: relative;
  background-color: white;
  margin: 5px;
  padding: 10px;
  border-radius: 5px;
  box-shadow: 2px 2px 5px rgba(0, 0, 0, 0.25);
  border: 4px solid currentColor;
  background-color: #2d3436;
}

.notification--status-notification {
  color: #74b9ff;
}

.notification--status-success {
  color: #00b894;
}

.notification--status-warning {
  color: #fdcb6e;
}

.notification--status-error {
  color: #d63031;
}

.notification span {
  display: block;
  padding-right: 60px;
  font-size: 0.8em;
  line-height: 1.2;
  font-weight: 100;
}

.notification button {
  /* reset */
  border: none;
  margin: 0;
  padding: 0;
  width: auto;
  overflow: visible;
  background: transparent;
  color: inherit;
  font: inherit;
  line-height: normal;
  -webkit-font-smoothing: inherit;
  -moz-osx-font-smoothing: inherit;
  -webkit-appearance: none;

  position: absolute;
  right: 0;
  top: 50%;
  transform: translateY(-50%);
  font-size: 0.6em;
  padding: 10px;
  width: 60px;
}

.notification button:hover {
  cursor: pointer;
  text-decoration: underline;
}
