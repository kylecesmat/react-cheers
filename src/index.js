// @flow
import * as React from 'react';

export type Notification = {
  kind: 'warning' | 'notification' | 'error' | 'success',
  message: string,
  id: string | number,
  expires?: number // ms
};

export type NotificationContext = {
  notifications: Array<Notification>,
  dispatch: Function,
  expire: Function
};

// typedef pending https://github.com/facebook/flow/pull/5920
export const Context = React.createContext({
  notifications: [],
  dispatch: () => {},
  expire: () => {}
});

type Props = {
  children: React.Node,
  defaultExpires?: number, // ms
  defaultKind?: 'warning' | 'notification' | 'error' | 'success'
};

type State = NotificationContext;

export class Provider extends React.Component<Props, State> {
  constructor() {
    super();

    this.state = {
      notifications: [],
      dispatch: this.dispatch,
      expire: this.expire
    };
  }

  // eslint-disable-next-line consistent-return
  dispatch = ({ id, message, kind, expires }: Notification) => {
    if (this.state.notifications.find((notif) => notif.id === id)) return null;

    if (!id) id = new Date().getUTCMilliseconds();

    const notificationExpires = expires || this.props.defaultExpires;
    if (notificationExpires) {
      setTimeout(() => this.expire(id), notificationExpires);
    }

    const update = [...this.state.notifications].concat([
      ({
        id,
        message,
        expires: notificationExpires,
        kind: kind || this.props.defaultKind || 'notification'
      }: Notification)
    ]);

    this.setState({ notifications: update });
  };

  expire = (id: string | number) => {
    const update = [...this.state.notifications].filter(
      (notif) => notif.id !== id
    );

    this.setState({ notifications: update });
  };

  render() {
    return (
      <Context.Provider value={this.state}>
        {this.props.children}
      </Context.Provider>
    );
  }
}

export const withNotifications = (Component: React$Element<*>) => (
  props: *
) => (
  <Context.Consumer>
    {(notificationContext) => <Component {...notificationContext} {...props} />}
  </Context.Consumer>
);

export const Composer = () => (
  <div className="notification-composer">
    <Context.Consumer>
      {({ notifications, expire }) =>
        notifications.map(({ kind, message, id }) => (
          <div
            key={id}
            className={['notification', `notification--status-${kind}`].join(
              ' '
            )}
          >
            <span>{message}</span>
            {expire && <button onClick={() => expire(id)}>Dismiss</button>}
          </div>
        ))
      }
    </Context.Consumer>
  </div>
);
