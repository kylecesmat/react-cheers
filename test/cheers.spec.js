import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from '../src';

describe('Provider', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<Provider>Child</Provider>);

    expect(wrapper).toMatchSnapshot();
  });
});
