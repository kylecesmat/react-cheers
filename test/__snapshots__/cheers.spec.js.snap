// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider should match snapshot 1`] = `
<[object Object]
  value={
    Object {
      "dispatch": [Function],
      "expire": [Function],
      "notifications": Array [],
    }
  }
>
  Child
</[object Object]>
`;
